src/dotnet-core-uninstall/Shared/VSVersioning/VisualStudioSafeVersionsExtractor.cs

```diff
@@ -2,51 +2,93 @@
     internal static class VisualStudioSafeVersionsExtractor
     {
-        // Divisions within version bands. Not inclusive, groupings: [2.1.000, 2.1.600) [2.1.600, 2.2.00) [2.0.00, 2.2.200) [2.2.200, 2.3.00)
-        private static readonly SemanticVersion[] SpecialCaseDivisions = { new SemanticVersion(2, 1, 600), new SemanticVersion(2, 2, 200) };
+        // The tool should not be used to uninstall any more recent versions of the sdk
+        public static readonly SemanticVersion UpperLimit = new SemanticVersion(5, 0, 0);
+
+        // Must keep one of each of these divisions to ensure Visual Studio works. 
+        // Pairs are [inclusive, exclusive)
+        private static readonly Dictionary<(SemanticVersion, SemanticVersion), string> VersionDivisionsToExplaination = new Dictionary<(SemanticVersion, SemanticVersion), string>
+        {
+            { (new SemanticVersion(1, 0, 0), new SemanticVersion(2, 0, 0)),  string.Format(LocalizableStrings.RequirementExplainationString, "") },
+            { (new SemanticVersion(2, 0, 0), new SemanticVersion(2, 1, 300)), string.Format(LocalizableStrings.RequirementExplainationString, "") },
+            { (new SemanticVersion(2, 1, 300), new SemanticVersion(2, 1, 600)), string.Format(LocalizableStrings.RequirementExplainationString, " 2017") },
+            { (new SemanticVersion(2, 1, 600), new SemanticVersion(2, 1, 900)), string.Format(LocalizableStrings.RequirementExplainationString, " 2019") },
+            { (new SemanticVersion(2, 2, 100), new SemanticVersion(2, 2, 200)), string.Format(LocalizableStrings.RequirementExplainationString, " 2017") },
+            { (new SemanticVersion(2, 2, 200), new SemanticVersion(2, 2, 500)), string.Format(LocalizableStrings.RequirementExplainationString, " 2019") },
+            { (new SemanticVersion(2, 2, 500), UpperLimit), string.Format(LocalizableStrings.RequirementExplainationString, "") }
+        };
+
+        private static (IDictionary<IEnumerable<Bundle>, string>, IEnumerable<Bundle>) ApplyVersionDivisions(IEnumerable<Bundle> bundleList)
+        {
+            var dividedBundles = new Dictionary<IEnumerable<Bundle>, string>();
+            foreach (var (division, explaination) in VersionDivisionsToExplaination)
+            {
+                var bundlesInRange = bundleList.Where(bundle => bundle.Version is SdkVersion && division.Item1 <= bundle.Version.SemVer && bundle.Version.SemVer < division.Item2);
+                bundleList = bundleList.Except(bundlesInRange);
+                if (bundlesInRange.Count() > 0)
+                {
+                    dividedBundles.Add(bundlesInRange, explaination);
+                }
+            }
+
+            return (dividedBundles, bundleList);
+        }
 
         public static IEnumerable<Bundle> GetUninstallableBundles(IEnumerable<Bundle> bundles)
         {
+            if (!RuntimeInfo.RunningOnWindows) // TODO this will be changed when mac protection is added
```

Comment:
this is enough for mac protection right? can we remove it now?

in general no TODO in code, you can file an issue and link the code there

Reply:
also, if you feel you really need to add a note to tell where to add the next feature, maybe it is an indication to refactor the code to make it more modular so the place to change is obvious.